### ISSUE TITLE
create an Addon for Cloud9

### ISSUE THREAD

**Issue by jstrachan** [jenkins-x/jx#776]
this looks an interesting start
https://github.com/errordeveloper/k9c

wonder if we could align the workspaces and [DevPods](https://jenkins-x.io/developing/devpods/)?

**Comment by errordeveloper** [comment]
Right now k9c just packages cloud9 in a container along with `kubeclt`, `docker` and `git`.   It mostly works fine for our use, I haven't seen any major issues. I'm happy to refactor any of the bits, or simply let you copy whatever you want out of k9c project (it'd be mostly just what's under `build/`).

Few things to note are:

- Cloud9 installer is a mess and k9c tries to keep things tight
- Cloud9 project doesn't actually have releases
- It required basic auth, which mostly of works for our current use where each workshop attendee gets a random password, but disabling auth would need a little more work

Few thoughts about the terminal feature:

- It uses `tmux` for the terminal (which is kind of nice, as you can reload the page and see the same thing in your terminal)
- It's not clear if it'd be possible to point terminal at another container, not the one where Cloud9 node app runs (it'd be nice to not put all eggs in one basket and run it as isolated side-car – maybe some clever tmux config could do the trick)

More thoughts:

- It's not clear to me how to customise the initial page you see
- Run button at the top is kind of pointless, it'd be nice to figure out how 
- There some UI features that point the user at hosted AWS version, it'd be nice to strip those out

Broadly, I think it's the least of all evils and is quite usable, especially for training.

**Comment by jstrachan** [comment]
@errordeveloper do you think it'll make things easier if there was a single container for the DevPod for developer tools (e.g.` go/maven/swift/ruby/whatever` - plus `skaffold, helm, kubectl, git` etc) - then a sidecar for Cloud9; or use some kind of initContainer to copy/install Cloud9 into a DevPod without having to create parallel devpod container images for each language + each language+cloud9?

Basically I'd like the Terminal inside the Cloud9 IDE to have all the stuff from the various different build images we use for each language/runtime tools

**Comment by jstrachan** [comment]
@errordeveloper looking at your `Dockerfile` it doesn't look too crazy hard to install cloud9 into the single DevPod container - though its extra complex as different build pods use different base images so we'd need to deal with `yum` and `apt-get` etc. So we could maybe install cloud9 dynamically in a DevPod if we have to - a little icky - but possible. 

Though the cleanest solution would be keeping cloud9 in a separate container really - if we could figure out how to get the Terminal to use the other container (with all the tools inside) - no idea how easy/hard thats gonna be; I guess its just changing the terminal command from `bash` to be `kubectl exec -it $POD bash` or something...

**Comment by errordeveloper** [comment]
I would rather see it as sidecar, although initial implementation would be
easier without that requirement.

Cloud9 depends on an old version of Node.js (it might work on a newwever
one, but according to the docs it wants v4. So you would limit yourself to
that version of Node.js.

**Comment by jstrachan** [comment]
@errordeveloper ah thanks. Agree - a sidecar would be simpler! Shame cloud9 isn't written in go so we could just statically link a binary and be done with it ;)

I wonder how easy it'd be to tweak cloud9 to allow a sidecar to be used to run all the terminals / builds / compiles etc?

**Comment by errordeveloper** [comment]
On Wed, 16 May 2018, 8:45 am James Strachan, <notifications@github.com>
wrote:

> @errordeveloper <https://github.com/errordeveloper> looking at your
> Dockerfile it doesn't look too crazy hard to install cloud9 into the
> single DevPod container - though its extra complex as different build pods
> use different base images so we'd need to deal with yum and apt-get etc.
> So we could maybe install cloud9 dynamically in a DevPod if we have to - a
> little icky - but possible.
>
Well, you could try using Cloud9 official install script, which supposed to
support different versions of Linux, but it tries to do many things, it is
kind of begging for some containers :)

https://github.com/c9/install/blob/master/install.sh

> Though the cleanest solution would be keeping cloud9 in a separate
> container really - if we could figure out how to get the Terminal to use
> the other container (with all the tools inside) - no idea how easy/hard
> thats gonna be; I guess its just changing the terminal command from bash
> to be kubectl exec -it $POD bash or something...
>
I think a reasonable way to do this would be via tweaking tmux config, set
shell there and no need to override the system shell (e.g. in case Cloud9
decides to run some random commands, it still wants to see an ordinary
local shell).

**Comment by errordeveloper** [comment]
I would focus on making terminal work, I think tmux should make it pretty
straightforward.

I have no idea how builds are supposed to work, maybe the extreme would be
to patch the whatever "plugin" implememts builds, but providing shims for
supported build tools maybe worth a try first.

**Comment by jstrachan** [comment]
@errordeveloper many thanks! 

I'm a total tmux noob - but I guess we could try hack tmux so that any new session really creates a shell in the other container than the IDE right?

**Comment by jstrachan** [comment]
@errordeveloper ooh - adding this file into `/root/.profile` seems to do the trick!!!
```
kubectl exec -it -n jx  $HOSTNAME bash
```

**Comment by jstrachan** [comment]
it kinda worked; though I was a little worried cloud9 was not being released frequently. I'm secretly hoping with the Microsoft <-> GitHub deal happening that we'll get a web based VS Code at some point! So was hoping something like that became a thing soon!

**Comment by jstrachan** [comment]
gonna close this - as we've Theia which is working great so far! https://jenkins-x.io/developing/devpods/#using-theia-ide
